enable departure time in josev

Abby-Wheelis opened the pull request:
## Describe your changes
These patches enable departure time simulation in this library

## Issue ticket number and link
[#220](https://github.com/EVerest/EVerest/issues/220)

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

replaces #40 - issues with signing, rebase is failing, and excessive churn

Abby-Wheelis committed:
enable departure time in josev

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas commented on iso15118/evcc/controller/simulator.py:
```diff
@@ -610,13 +616,28 @@ async def process_sa_schedules_v2(
                 )
                 evcc_profile_entry_list.append(last_profile_entry_details)
 
+                # Set Curve Variables...
```
Align the comment to the left

Abby-Wheelis replied:
[67d83c0](https://github.com/EVerest/ext-switchev-iso15118/pull/45/commits/67d83c055c29e5e0c77f6279caabe699649c42c9)

SebaLukas commented on iso15118/evcc/states/iso15118_2_states.py:
```diff
@@ -9,6 +9,7 @@
 from typing import Any, List, Union
 import os
 
+from iso15118.evcc.everest import context as EVEREST_CONTEXT
```
context is already imported in line 120. Please delete this line

Abby-Wheelis replied:
[80b840a](https://github.com/EVerest/ext-switchev-iso15118/pull/45/commits/80b840a64b551c8523c3e70d924b70d542c181d8)

SebaLukas commented on iso15118/evcc/states/iso15118_2_states.py:
```diff
@@ -113,6 +114,8 @@
 
 logger = logging.getLogger(__name__)
 
+EVEREST_EV_STATE = EVEREST_CONTEXT.ev_state
```
Move to line 120 and use `EVEREST_CTX` instead of `EVEREST_CONTEXT`

Abby-Wheelis replied:
[80b840a](https://github.com/EVerest/ext-switchev-iso15118/pull/45/commits/80b840a64b551c8523c3e70d924b70d542c181d8)

SebaLukas commented on iso15118/evcc/states/iso15118_2_states.py:
```diff
@@ -1170,6 +1185,13 @@ async def process_message(
         # EVerest code start #
         if charging_status_res.evse_max_current:
             evse_max_current = charging_status_res.evse_max_current.value * pow(10, charging_status_res.evse_max_current.multiplier)
+            EVEREST_CTX.publish('AC_EVSEMaxCurrent', evse_max_current)
```
Remove this line. `evse_max_current` is already published in line 1195

Abby-Wheelis replied:
[61120ab](https://github.com/EVerest/ext-switchev-iso15118/pull/45/commits/61120ab332dc2339dbe226cf862b983a262d2e5a)

SebaLukas commented on iso15118/evcc/states/iso15118_2_states.py:
```diff
@@ -798,11 +804,20 @@ async def process_message(
+                self.comm_session.end_of_profile_schedule = 86400
+
             EVEREST_CTX.publish('ev_power_ready', True)
+            EVEREST_CTX.publish('AC_EVPowerReady', True)
```
Remove this line

Abby-Wheelis replied:
[597ce71](https://github.com/EVerest/ext-switchev-iso15118/pull/45/commits/597ce716aa502f564332b8790a192c5fd1c38708)

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,115 @@
+ Check notice on line 38 in iso15118/evcc/states/power_curve.py
+
+
+Codacy Production
+/ Codacy Static Code Analysis
+iso15118/evcc/states/power_curve.py#L38
```
Remove these lines

Abby-Wheelis replied:
[9a11a1f](https://github.com/EVerest/ext-switchev-iso15118/pull/45/commits/9a11a1f40765c93133e3438a5cccae3572d61687)

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,115 @@
+    curve_schedule = convert_tuple_schedule(curve_schedule)
+    logger.debug(f"Returning a curve schedule of: {curve_schedule}")
+
+    return curve_schedule
```
Add an extra line here

Abby-Wheelis replied:
[9179c98](https://github.com/EVerest/ext-switchev-iso15118/pull/45/commits/9179c9897f6dd87fe4dc1ace87c8bfa44c11af15)

Abby-Wheelis committed:
Fix indentation

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
Clean up context import

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
Remove extra publish

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
Remove line

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove extra comment

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
Newline

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
Merge remote-tracking branch 'upstream/everest' into dt_signed_final

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas commented on iso15118/evcc/controller/simulator.py:
```diff
@@ -294,7 +299,7 @@ async def get_charge_params_v2(self, protocol: Protocol) -> ChargeParamsV2:
 
         if (await self.get_energy_transfer_mode(protocol)).startswith("AC"):
-            e_amount = PVEAmount(multiplier=0, value=60,
+            e_amount = PVEAmount(multiplier=3, value=EVEREST_EV_STATE.EAmount,
```
Can you guarantee that `EVEREST_EV_STATE.EAmount` is in `kWh` and not in `Wh`?

Abby-Wheelis replied:
I'm not sure that I can guarantee that. I've set all the defaults in kWh, but we're allowing for user input, so someone could mess up and put in a value in Wh. Is there an expected range that is allowed? We could check inputs for being inside of the expected range.

Abby-Wheelis replied:
Could checks like that be a follow-on or future fix?

Abby-Wheelis replied:
If someone puts in a number that is "too high" (such as mistakenly Wh), they would just get the maximum charge rate.

SebaLukas replied:
There is 2-3 options right now:
1. Change `EAmount` to `EAmount_kWh`. Then it is obvious for Josev and PyEvJosev.
2. Use this function to calcute the `RationalNumber`/`PVEAmount`:  https://github.com/EVerest/ext-switchev-iso15118/blob/everest/iso15118/secc/everest/__init__.py#L7 But in this case `EVEREST_EV_STATE.EAmount` should be in Wh.
3. Leave it as it is, but with the risk that incorrect EAmount values may be provided in the future. I would not prefer this option.

Abby-Wheelis replied:
>Use this function to calcute the RationalNumber/PVEAmount: https://github.com/EVerest/ext-switchev-iso15118/blob/everest/iso15118/secc/everest/__init__.py#L7 But in this case EVEREST_EV_STATE.EAmount should be in Wh.

I think this might be the way we should proceed, given that the other uses of EAmount in the repo seem to already work this way. EVEREST_EV_STATE will still be in kWh, but will be converted to Wh for use. 

I also completed 1) so that it was more clear that the value in the state was in kWh

SebaLukas reviewed (approved):
See my last comment

Abby-Wheelis committed:
PVEAmount for new departure time curve

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
clarify that the value in the state is in kWh

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,107 @@
+    deptime_linear is used to generate a linear power curve
+    for a given departure time, energy amount, and maximum power.
+'''
+def deptime_linear(DepTime: float, EAmount: float, PMax: float) -> tuple[list,list,list]:
```
Now `EAmount` has the type `PVEAmount` and not `float` anymore

Abby-Wheelis committed:
prevent missing variable

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
account for new type

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix curve calculations

update from kWh to W, account for format of lists
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
use departure time from the state

must use DT from the EV state, if no changes have been made to the state, the default of 7200 will be used.
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
value is a float, decimal value is retrieved and pased in when called
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
de-duplicate import
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
Merge remote-tracking branch 'upstream/everest' into dt_signed_final

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>
